### PR TITLE
feat(compiler): struct declarations

### PR DESCRIPTION
**Pull request by hasanaburayyan**
Closes: https://github.com/winglang/wing/issues/120

*By submitting this pull request, I confirm that my contribution is made under the terms of the [Monada Contribution License](https://docs.winglang.io/terms-and-policies/contribution-license.html)*.

### CODE REVIEW

File: libs/wingc/src/parser.rs [hunk 1]
```diff
@@ -332,6 +332,43 @@ impl<'s> Parser<'s> {
...
+					.expect("struct definition fields should always have a name"),
+			)?;
+			let type_ = field_node
+				.child_by_field_name("type")
+				.expect("struct definition fields should always have explicit type");
+			let f = ClassField {
+				name: identifier,
+				member_type: self.build_type_annotation(&type_)?,
+				reassignable: false,
+				phase: Phase::Preflight,
```
Comment (MarkMcCulloh) [inline libs/wingc/src/parser.rs:358]: Is it preflight? I feel like the fields themselves may be phase independent, not really sure though.

File: libs/wingc/src/type_check.rs [hunk 2]
```diff
@@ -836,6 +836,20 @@ impl TypeRef {
...
+	pub fn is_deep_mutable(&self) -> bool {
+		match &**self {
+			Type::MutArray(_) => true,
+			Type::MutMap(_) => true,
+			Type::MutSet(_) => true,
+			Type::MutJson => true,
+			Type::Array(v) => v.is_deep_mutable(),
+			Type::Map(v) => v.is_deep_mutable(),
+			Type::Set(v) => v.is_deep_mutable(),
+			_ => false,
```
Comment (MarkMcCulloh) [inline libs/wingc/src/type_check.rs:849]: Optionals?

Maybe it's worthwhile for this to be exhaustive anyways

File: libs/tree-sitter-wing/grammar.js [hunk 3]
```diff
@@ -108,9 +108,9 @@ module.exports = grammar({
       seq(
         "struct",
         field("name", $.identifier),
-        optional(seq("extends", commaSep($.identifier))),
+        optional(seq("extends", commaSep(field("extend", $.identifier)))),
```
Comment (MarkMcCulloh) [inline libs/tree-sitter-wing/grammar.js:111]: nit: I feel like this should be plural, both to match the keyword and also to better signify that there can be many of them.

```suggestion
        optional(seq("extends", commaSep(field("extends", $.identifier)))),
```

File: libs/wingc/src/parser.rs [hunk 4]
```diff
@@ -332,6 +332,43 @@ impl<'s> Parser<'s> {
 		})
 	}
 
+	fn build_struct_definition_statement(&self, statement_node: &Node) -> DiagnosticResult<StmtKind> {
+		let name = self.node_symbol(
+			&statement_node
+				.child_by_field_name("name")
+				.expect("struct definitions should always have a name"),
```
Comment (MarkMcCulloh) [inline libs/wingc/src/parser.rs:339]: nit: There's a helper method that can help avoid the expect

```suggestion
			&self.get_child_field(&statement_node, "name")?
```

**Review by MarkMcCulloh — APPROVED** [review]
I'm slightly concerned about which phase the fields should be, so maybe that needs to be double checked, but I otherwise think this looks great!

File: libs/wingc/src/type_check.rs (hunk 2 above)
Comment (hasanaburayyan) [reply libs/wingc/src/type_check.rs:849]: Great catch!

File: libs/wingc/src/parser.rs (hunk 4 above)
Comment (hasanaburayyan) [reply libs/wingc/src/parser.rs:339]: That is fucking wonderful! Thanks for the nit!

**Comment by Chriscbr** [comment]
Sweet!! 😄

**Pull request merged**